Crete concept page for user task listeners (#4807)
camunda/camunda-docs: docs/components/concepts/user-task-listeners.md

---
id: user-task-listeners
title: "User task listeners"

> description: "User task listeners allow users to react to specific user task lifecycle events by executing custom logic."

Reviewer: Speaking from a "beginner" perspective -- when you say custom logic do you mean logic rules for standard and custom objects by using decision tables or functions?

Author: Hi @christinaausley! 👋
Thank you for your feedback - it’s always valuable to have the beginner's perspective.  

To clarify, "custom logic" in this context refers to the ability for users to write their own Java code that can be executed in response to specific user task lifecycle events, such as `assigning` or `completing` a task. This allows users to implement custom behaviors and rules tailored to their specific use cases, such as triggering external systems, altering user task data using new `corrections` feature while completing user task listener job and so on.  

We've updated the description to make this more explicit and beginner-friendly:  
> User task listeners allow users to react to specific user task lifecycle events by executing custom Java code, enabling dynamic handling of user tasks.

Let me know if you think this adjustment improves clarity or if there's anything else you'd recommend! 😊

---

> A user task listener (UTL) allows users to react to specific user task lifecycle events by executing custom logic.

Reviewer: Is (UTL) necessary here? We still write out user task listener throughout the rest of the doc, and I don't think we reference UTL anywhere else in the docs, so I think this can be removed?

Author: Yeah, thank you for pointing that out - you're absolutely right. Since "UTL" isn’t used anywhere else within the user task listeners concept page, we’ll go ahead and remove it to maintain consistency and avoid unnecessary abbreviations.  

Appreciate the catch! 😊

## About user task listeners

> User task listeners enable flexibility and control over user task behavior by allowing external logic to react to lifecycle events. This is useful for implementing custom business logic, assigning tasks, or triggering external systems without cluttering the BPMN model. Additionally, user task data can be dynamically altered during execution using the corrections feature, enabling adjustments to key attributes such as assignee, due date, priority and so on, before task state transitions are finalized.

Author: Thank you for the suggestion, Nico! I’ve rephrased your proposal slightly to improve clarity and flow while keeping the focus on the functionalities you mentioned.
Here's the updated version:

```
User task listeners provide flexibility and control over user task behavior:

- They can react to user task lifecycle events, such as assigning and completing.
- They can access user task data, such as the assignee, to execute task-specific business logic.
- They can dynamically correct user task data during execution, allowing adjustments to key attributes such as the assignee, due date, and priority.
- They can deny state transitions, rolling back the task to its previous state, which enables validation of task lifecycle changes.
```

### Use cases

User task listeners are useful in the following scenarios:

- Implementing complex user task assignment or reassignment logic
- Validating user task assignments
- Notifying users of new task assignments with contextual information

> - Reacting to task completions with custom logic

Reviewer 2: 💭 (disclaimer: probably newbie question) what is a use-case to react to completions with custom logic based on events? I guess the same functionality could be achieved through usual BPMN flow, right?

> Reacting to task completions with custom logic

> - Reacting to task completions with custom logic

Reviewer 3: @eppdot Yes, to avoid cluttering the BPMN model with technical details. This is also a typical case for [Execution Listeners](https://docs.camunda.io/docs/next/components/concepts/execution-listeners/#about-execution-listeners).

Currently, user task listeners support the following events:

- **Assigning**: Triggered while assigning a user task.
- **Completing**: Triggered while completing a user task.

### User task listener properties

Each user task listener has the following properties:

| Property    | Description                                                                                                   |
| :---------- | :------------------------------------------------------------------------------------------------------------ |
| `eventType` | Specifies the user task event that triggers the listener. Supported values are `assigning` and `completing` . |
| `type`      | The name of the job type.                                                                                     |
| `retries`   | The number of retries for the user task listener job.                                                         |

:::note
If multiple user task listeners of the same `eventType` (such as multiple `assigning` listeners) are defined on the same user task, they are executed sequentially, one after the other, in the order they are defined in the BPMN model.
:::

## Implement a user task listener

> User task listeners are implemented using [job workers](/components/concepts/job-workers.md), similar to execution listeners and service task jobs. The job worker processes the task listener job, applies corrections, and may optionally deny the operation.

Reviewer 3: ```suggestion
User task listeners are implemented using [job workers](/components/concepts/job-workers.md), similar to execution listeners and service task jobs. The job worker processes the task listener job, can apply corrections, and may optionally deny the lifecycle transition.
```

See [open a job worker](/apis-tools/java-client-examples/job-worker-open.md) for an example of how to create a job worker and handler that can also process user task listener jobs.

### Accessing user task data

> User task-specific data, such as `assignee` and `priority`, is accessible through the job headers of the user task listener job. The following properties can be retrieved using predefined header names from the `Protocol` class:
> 
> | Property          | Header name                                       |
> | ----------------- | ------------------------------------------------- |
> | `action`          | `Protocol.USER_TASK_ACTION_HEADER_NAME`           |
> | `assignee`        | `Protocol.USER_TASK_ASSIGNEE_HEADER_NAME`         |
> | `candidateGroups` | `Protocol.USER_TASK_CANDIDATE_GROUPS_HEADER_NAME` |
> | `candidateUsers`  | `Protocol.USER_TASK_CANDIDATE_USERS_HEADER_NAME`  |
> | `dueDate`         | `Protocol.USER_TASK_DUE_DATE_HEADER_NAME`         |
> | `followUpDate`    | `Protocol.USER_TASK_FOLLOW_UP_DATE_HEADER_NAME`   |
> | `formKey`         | `Protocol.USER_TASK_FORM_KEY_HEADER_NAME`         |
> | `userTaskKey`     | `Protocol.USER_TASK_KEY_HEADER_NAME`              |
> | `priority`        | `Protocol.USER_TASK_PRIORITY_HEADER_NAME`         |

Reviewer 3: The `Procotol` class is Java specific. The concept page should be language agnostic.

```suggestion
User task-specific data, such as `assignee` and `priority`, is accessible through the job headers of the user task listener job. The following properties can be retrieved using reserved header names:

| Property          | Header name                        |
| ----------------- | ---------------------------------- |
| `action`          | `io.camunda.zeebe:action`          |
| `assignee`        | `io.camunda.zeebe:assignee`        |
| `candidateGroups` | `io.camunda.zeebe:candidateGroups` |
| `candidateUsers`  | `io.camunda.zeebe:candidateUsers`  |
| `dueDate`         | `io.camunda.zeebe:dueDate`         |
| `followUpDate`    | `io.camunda.zeebe:followUpDate`    |
| `formKey`         | `io.camunda.zeebe:formKey`         |
| `userTaskKey`     | `io.camunda.zeebe:userTaskKey`     |
| `priority`        | `io.camunda.zeebe:priority`        |
```

Author: Thank you, Nico, for pointing out the visibility of the `Protocol` class!
I've applied your suggestion to make the concept page language-agnostic.

> Below is an example of accessing the `assignee` value from the headers:

Reviewer 3: ```suggestion
Below is an example of accessing the `assignee` value from the headers in Java:
```

```java

> import io.camunda.zeebe.protocol.Protocol;
> 
> final JobHandler userTaskListenerHandler =
>     (jobClient, job) -> {
>         // Access the 'assignee' from the job headers
>         // highlight-start
>         final String assignee = job.getCustomHeaders()
>             .get(Protocol.USER_TASK_ASSIGNEE_HEADER_NAME);
>         // highlight-end
> 
>         System.out.println("The assignee for this user task is: " + assignee);
> 
>         // remaining job handler logic
>     };

Reviewer 3: This example is not accessible for users only depending on `camunda-client-java` (as it doesn't come with `zeebe-protocol` unless you depend on our bom)

```suggestion
final JobHandler userTaskListenerHandler =
    (jobClient, job) -> {
        // Access the 'assignee' from the job headers
        // highlight-start
        final String assignee = job.getCustomHeaders()
            .get("io.camunda.zeebe:assignee");
        // highlight-end

        System.out.println("The assignee for this user task is: " + assignee);

        // remaining job handler logic
    `};`
```

```

Each header provides user task metadata that can be leveraged to customize the behavior of the user task listener job. Use these headers to retrieve necessary information about the user task in your job handler implementation.

> ### Correcting user task data

Reviewer 3: ❌ We should mention that corrections do not trigger the `UPDATING` lifecycle event.

Author: Thank you for the suggestion, Nico! I've addressed it by mentioning at the end of this sentence: 

> These corrections are immediately available to any subsequent task listeners and are applied to the user task when the lifecycle transition is finalized, **without triggering the `UPDATING` lifecycle event.**

> User task listeners can correct user task data before the lifecycle transition is finalized. Corrections allow user task listeners to update specific attributes of the user task, such as the assignee, due date, follow-up date, candidate users, candidate groups, and priority. These corrections are applied after all task listeners for the current operation have successfully completed.

Reviewer 3: ❌ When the corrections are applied depends on the perspective. For subsequent listeners, it's available immediately. Also note that they aren't applied when the lifecycle transition is denied.

```suggestion
User task listeners can correct user task data before the lifecycle transition is finalized. Corrections allow user task listeners to update specific attributes of the user task, such as the assignee, due date, follow-up date, candidate users, candidate groups, and priority. These corrections are available to any subsequent task listeners immediately, and are applied to the user task when the lifecycle transition is finalized.
```

> If an operation is denied by a listener, no corrections are applied to the user task.

Reviewer 3: ```suggestion
If a lifecycle transition is denied by a listener, no corrections are applied to the user task.
```

> Below is an example of how to correct the user task data form a job worker while completing the user task listener job:

Reviewer 3: ```suggestion
Below is an example of how to correct the user task data form a job worker while completing the user task listener job in Java:
```

```java
final JobHandler completeTaskListenerJobWithCorrectionsHandler =
    (jobClient, job) ->
        jobClient
            .newCompleteCommand(job)
            // highlight-start
            .withResult(
                new CompleteJobResult()
                    .correctAssignee("john_doe") // assigns the user task to 'john_doe'
                    .correctDueDate(null) // preserves the current 'dueDate' of the user task
                    .correctFollowUpDate("") // clears the 'followUpDate'
                    .correctCandidateUsers(List.of("alice", "bob")) // sets candidate users
                    .correctCandidateGroups(List.of()) // clears the candidate groups
                    .correctPriority(80)) // sets the priority to 80
            // highlight-end
            .send();

client.newWorker()
    .jobType("user-task-listener-completion") // type of the user task listener job
    .handler(completeTaskListenerJobWithCorrectionsHandler)
    .open();
```

> ### Denying the operation
> 
> User task listeners can deny a user task operation, such as assignment or completion, effectively preventing the lifecycle transition from completing.
> 
> When an operation is denied:
> 
> - **Corrections discarded**: Any corrections made by preceding listeners within the same operation are discarded.
> - **Task state preserved**: The user task retains its state and data as if the operation was never attempted.
> 
> This capability is particularly useful for implementing validation logic or enforcing business rules before allowing a user task operation to proceed.

Reviewer 3: :x: It's possible that the listener denies a lifecycle transition that isn't an operation (e.g. assigning on creation, or creation itself). To ensure correct language, we can discuss the lifecycle event (the event that the listener reacts on) and the lifecycle transition (the change which is often an operation but can also be initiated by the engine rather than an outside request).

```suggestion
### Denying the lifecycle transition

User task listeners can deny the user task lifecycle transition belonging to the lifecycle event. For example, it can deny the completion of a task in reaction to the completing event, effectively preventing a user request to complete the task.

When a lifecycle transition is denied:

- **Corrections discarded**: Any corrections made by preceding listeners within the same operation are discarded.
- **Task state preserved**: The user task retains its state and data as if the lifecycle event never occurred.

This capability is particularly useful for implementing validation logic or enforcing business rules before allowing a user task lifecycle transition to proceed.
```

Author: Thank you, Nico, for pointing this out! I've updated the section to use the proposed terminology by adopting your suggestion.

> Below is an example of how to deny a user task operation from a job worker while completing the user task listener job:

Reviewer 3: ```suggestion
Below is an example of how to deny a user task operation from a job worker while completing the user task listener job in Java:
```

```java
final JobHandler denyUserTaskOperationHandler =
    (jobClient, job) ->
        jobClient
            .newCompleteCommand(job)
            // highlight-start
            .withResult()
            .deny(true)
            // highlight-end
            .send();
```

## Expression evaluation and incident behavior

### Expression evaluation

> User task listener properties, such as job `type` or `retries`, are evaluated right before the job creation for the listener. If an expression evaluation fails, the operation is paused, and an incident is raised.

Reviewer: perhaps link to https://docs.camunda.io/docs/components/concepts/incidents/ here?

Reviewer: Although I see it is listed in related resources. Up to you.

### Incident recovery

> If a user task listener job fails or its expression evaluation raises an incident, the operation is paused until the incident is resolved.

Reviewer 3: ```suggestion
If a user task listener job fails or its expression evaluation raises an incident, the lifecycle transition is paused until the incident is resolved.
```

There are two types of incidents for task listeners:

> - **Expression evaluation failure**: Raised when a property expression (e.g., `type`) fails to evaluate.

Reviewer: ```suggestion
- **Expression evaluation failure**: Raised when a property expression (for example, `type`) fails to evaluate.
```

- **Job failure**: Raised when a user task listener job exhausts all retries.

## Limitations

User task listeners have the following limitations:

- **Limited events support**: Currently, only `assigning` and `completing` events are supported.
- **No variable handling**: User task listener jobs cannot be completed if variables are provided.
- **No BPMN error throwing**: Throwing BPMN errors from user task listener jobs is not supported.

## Related resources

- [Job workers (basics)](/components/concepts/job-workers.md)
- [Job workers (Java client)](/apis-tools/java-client/job-worker.md)
- [Incidents](/components/concepts/incidents.md)
- [Expressions](/components/concepts/expressions.md)

> - [Execution listeners](/components/concepts/execution-listeners.md)

Author: Yeah, good point! I've added a reference to the `User tasks` page in the `Related resources` section.
